hw/ip/otbn/dv/rig/rig/gens/bad_bnmovr.py

```diff
@@ -0,0 +1,96 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import random
+from typing import Optional
+
+from shared.operand import RegOperandType, OptionOperandType
+from shared.insn_yaml import InsnsFile
+
+from ..config import Config
+from ..model import Model
+from ..program import ProgInsn, Program
+
+from ..snippet import ProgSnippet
+from ..snippet_gen import GenCont, GenRet, SnippetGen
+
+
+class BadBNMovr(SnippetGen):
+    '''A snippet generator that generates program ending BN.MOVR instructions.
+
+    This includes incrementing both GPRs or having *grd or *grs > 31
+
+    '''
+
+    ends_program = True
+
+    def __init__(self, cfg: Config, insns_file: InsnsFile) -> None:
+        super().__init__()
+
+        self.insn = self._get_named_insn(insns_file, 'bn.movr')
+
+        # bn.movr expects the operands: grd, grs, grd_inc, grs_inc
+        if len(self.insn.operands) != 4:
+            raise RuntimeError('Unexpected number of operands for bn.movr')
+
+        grd, grs, grd_inc, grs_inc = self.insn.operands
+        exp_shape = (
+            # grd
+            isinstance(grd.op_type, RegOperandType) and
+            grd.op_type.reg_type == 'gpr' and
+            not grd.op_type.is_dest() and
+            # grs
+            isinstance(grs.op_type, RegOperandType) and
+            grs.op_type.reg_type == 'gpr' and
+            not grs.op_type.is_dest() and
+            # grd_inc
+            isinstance(grd_inc.op_type, OptionOperandType) and
+            # grs_inc
+            isinstance(grs_inc.op_type, OptionOperandType)
+        )
+        if not exp_shape:
+            raise RuntimeError('Unexpected shape for bn.movr')
+
+        self.weight = cfg.insn_weights.get(self.insn.mnemonic)
+
+        # Check that the instruction has a positive weight
+        if not self.weight:
+            self.disabled = True
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        # Get known registers
+        known_regs = model.regs_with_known_vals('gpr')
+        base = []
+
+        # Initialize operand values
+        reg_val_grd = 0
+        reg_val_grs = 0
+
+        # Make sure there is at least one out of bounds value
+        for reg_idx, reg_val in known_regs:
+            if (31 < reg_val):
+                base.append((reg_idx, reg_val))
+
+        if not base:
+            return None
+
+        # Even if one of them is unacceptable, that will be enough.
+        while not (reg_val_grd > 31 or reg_val_grs > 31):
+            # Pick a random register among known registers that is constrained
+            op_val_grd, reg_val_grd = random.choice(known_regs)
+            # We don't care about the actual value, only interested in index
+            op_val_grs, reg_val_grs = random.choice(known_regs)
+
+        op_val = [op_val_grd, op_val_grs, 1, 1]
```

Comment:
I think this will only produce the "both increments" case when one of the GPRs is >= 32. Could you rejig things a bit so that we can also see the situation with both increments but valid GPRs?

Reply:
rejigged! :-)